[Tooling] Split Fastfile for readability and move lanes around

AliSoftware opened the pull request:
This is a follow-up on https://github.com/wordpress-mobile/WordPress-iOS/pull/18061 — see idea 2 of the "Next Steps" section

# What is it about?

This reorganizes the lanes in `Fastfile` / `Jetpack-Fastfile` / `ScreenshotFastfile` into a different kind of split, more topic-oriented, to facilitate maintenance.

I ended up with the following `lanes/*.rb` "topics": `build`, `codesign`, `localization`, `release` and `screenshots`

It also does some `rubocop` cleanup on the code that has been extracted, and make the documentation for each lane less verbose and more readable (and idiomatic to ruby doc)

# PR Review Guidance

## Commits have been kept independant

I've tried to keep each commit focused for easier review.
 - First set of commits is mostly cut/pasting the code from `Fastfile` to `lanes/*.rb` files without being changed
 - Next set of commits change things like lane comments
 - Last set of commits apply `rubocop --auto-correct` for each style rule it detected violations for.

## Use this script to review the **_implementation_** changes

Since it might not be easy to review the PR's diff itself to ensure that the logic of the implementation of each lane have not changed at all (but only things like `"` -> `'` quotes by rubocop) — and validate that every `lane` and `def` blocks have only been _moved_, not _modified_ — I've created a small script to make the review easier:

<details><summary><tt>diff_lanes.rb</tt></summary>

```ruby
#!/usr/bin/env ruby

require 'fileutils'

OUTPUT_DIR = 'fastlane-lanes-diff-report'

def parse_lanes(content)
  current_lane = nil
  lane_indentation = ''
  buffer = ''
  implementations = {}

  content.lines.each do |line|
    # Detect `lane` or `def`, remember name and indentation, and start buffering next lines (aka implementation)
    lane_match = /^(\s*)lane (:\w*) +do/.match(line)&.captures || /^(\s*)def (\w*)/.match(line)&.captures
    if lane_match
      current_lane = lane_match[1]
      lane_indentation = lane_match[0]
    end

    next if current_lane.nil? # Skip any line outside a `lane` or `def` block

    buffer += line

    # Detect `end` keyword with matching indentation
    if /^#{lane_indentation}end$/.match?(line.chomp)
      implementations[current_lane] = buffer
      current_lane = nil
      lane_indentation = ''
      buffer = ''
    end
  end

  implementations
end

# Construct file path for given lane and suffix
def fpath(lane, suffix)
  (type, name) = lane.start_with?(':') ? ['lane', lane.delete_prefix(':')] : ['def', lane]
  File.join(OUTPUT_DIR, "#{type}-#{name}-#{suffix}.txt")
end

# Dump implementation of each lane/def in a dedicated `.txt` file. Raise if file already exist (suggesting the lane/def was declared/detected twice)
def dump_lanes(h, suffix:)
  h.each do |lane, impl|
    file = fpath(lane, suffix)
    raise "File already exists: #{file}" if File.exist?(file)
    File.write(file, impl)
  end
end


### MAIN ###

raise "Please delete #{OUTPUT_DIR} before starting" if Dir.exist?(OUTPUT_DIR) && !Dir.empty?(OUTPUT_DIR)
FileUtils.mkdir_p(OUTPUT_DIR)

# Parse **new** lanes from `Fastfile` + `fastlane/lanes/*.rb`
(['fastlane/Fastfile'] + Dir.glob('fastlane/lanes/*.rb')).each do |f|
  puts "--- Parsing: #{f}..."
  new_lanes = parse_lanes(File.read(f))
  dump_lanes(new_lanes, suffix: 'new')
end

# Parse **old** lanes from trunk's `Fastfile`, `Jetpack-Fastfile` and `ScreenshotFastfile`
puts "--- Parsing lanes from trunk..."
trunk_lanes = {}.merge(
  parse_lanes(`git show trunk:fastlane/Fastfile`),
  parse_lanes(`git show trunk:fastlane/Jetpack-Fastfile`),
  parse_lanes(`git show trunk:fastlane/ScreenshotFastfile`)
)
dump_lanes(trunk_lanes, suffix: 'trunk')

# Compute diffs between the two
trunk_lanes.each do |lane, impl|
  diff = `diff -w "#{fpath(lane, 'trunk')}" #{fpath(lane,'new')}`
  File.write(fpath(lane, 'diff'), diff)
end
```
</details>

Save the script above in a `diff_lanes.rb` file at root of the repo, be sure you're `git checkout` on this branch, then run the script (`ruby diff_lanes.rb`). This will:

 - Find all the `def` and `lane` declarations in the **old** `Fastfile`/`Jetpack-Fastfile`/`ScreenshotFastfile` files currently in `trunk`, then create a `<def|lane>-<name>-trunk.txt` file for each lane/def found there, with the lane/def implementation as its content
 - Then it will do the same for all the `def`/`lane` found in the **new** `Fastfile` + `lanes/*.rb` files, and create `<def|lane>-<name>-new.txt`
 - Finally, for each lane/def, it will `diff -w` the 2 corresponding files to compare the old vs new implementations, and dump the diff in `<def|lane>-<name>-diff.txt`

You can then easily review especially the `*-diff.txt` files to ensure that the only changes in the implementation of each lane — before/after the cut/paste from their old location to the new `lanes/*.rb` — are small changes from `rubocop` (like quotes changed from `"` to `'` and the likes) or similar things that make sense but don't really change the implementation content.

Note: The script is expected to also print an error `diff: fastlane-lanes-diff-report/def-ensure_sentry_installed-new.txt: No such file or directory` when run. This is actually expected, because I indeed deleted the `def ensure_sentry_installed` method (to use the `sentry_check_cli_installed` action directly) so while that `def` was present in `trunk`, it no longer exist in this branch.

## To test

 - Run `fastlane lanes` and validate that it still prints the same list of lanes
 - Run `rubocop fastlane/**/*.rb fastlane/Fastfile` and validate that there's no violation reported
 - Using the script provided in the previous section, compare the implementation of each lane/function before vs after and validate that the little small changes are only stylistic (`rubocop --auto-correct`) and all make sense
 - Review the new, shorter code-documentation and `desc` of each lane to validate it's clear, no typos, and correct.

AliSoftware committed:
Move around constants and imports to group things

AliSoftware committed:
Move all L10n lanes to lanes/localization.rb

AliSoftware committed:
Move build-related lanes to lanes/build.rb

AliSoftware committed:
Move signing-related lanes to lanes/codesign.rb

AliSoftware committed:
Move test lane to lanes/build.rb too

AliSoftware committed:
Move release-related lanes to lanes/release.rb

AliSoftware committed:
Move `UPLOAD_TO_APP_STORE_COMMON_PARAMS` constant

Closer to the lanes it's used with, aka in lanes/localization.rb

AliSoftware committed:
Move `trigger_buildkite_release_build` helper function

Closer to the lanes it's used with, aka in lanes/release.rb

AliSoftware committed:
Move `gutenberg_dep_check` helper lane

To lanes/release.rb, closer to where it's used

AliSoftware committed:
Move release notes helpers to lanes/release.rb

AliSoftware committed:
Remove `ensure_sentry_installed` helper

And replace it with direct calls to sentry_check_cli_installed to avoid the need for a trampoline helper.

AliSoftware committed:
Move `generate_installable_build_number` helper

To lanes/build.rb so it's closer to the lanes using it

AliSoftware committed:
Move `read_version_from_config`

AliSoftware committed:
Reduce noise in lanes/release.rb doc

AliSoftware commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,225 @@
+    print_release_notes_reminder
+  end
+
+  # Completes the final steps for the code freeze
+  #  - Generate `.strings` files from code then merge the other, manually-maintained `.strings` files with it
+  #  - Trigger the build of the first beta on CI
+  #
+  # @option [Boolean] skip_confirm (default: false) If true, avoids any interactive prompt
+  #
+  desc 'Completes the final steps for the code freeze'
+  lane :complete_code_freeze do |options|
```
@mokagio Can I get a preliminary review (even if PR is still draft) and feedback on the direction I'm going towards here, especially w.r.t the new, less verbose and less visually-noisy lanes' documentation comments

---

Other ideas I've tried include:

<details><summary> 1️⃣  Moving all the docs into the <tt>desc</tt> string</summary>

After all, it's supposed to be what `desc` is for… 🤔 

```ruby
  desc <<~DESC
    Completes the final steps for the code freeze
     - Generate `.strings` files from code then merge the other, manually-maintained `.strings` files with it
     - Trigger the build of the first beta on CI
     
      @option [Boolean] skip_confirm (default: false) If true, avoids any interactive prompt
  DESC
  lane :complete_code_freeze do |options|
```

 - Which has the benefit of showing the whole help in the console when you run `fastlane lanes` to list them all
 - But has the big drawback imho that the documentation is then seen as a regular `String` by your IDE's syntax highlighter (and not a comment anymore), and thus will _not_ pop up as nicely nor be as easily differentiated by color as when it is written as a `# comment`. Which is why I prefer keeping that doc as a `# comment` after all.

</details>

<details><summary>2️⃣  Mixing <tt>desc</tt> tagline and the more-detailed doc-comments</summary>

As I was not _super_ happy with the `desc` often being a repetition of the first line of the doc-comment (though it's not _always_ the case, e.g. sometimes I've used a `desc` which is more concise), I thought I'd try avoiding that repetition using something like below:

```ruby
  desc 'Completes the final steps for the code freeze'
  #  - Generate `.strings` files from code then merge the other, manually-maintained `.strings` files with it
  #  - Trigger the build of the first beta on CI
  #
  # @option [Boolean] skip_confirm (default: false) If true, avoids any interactive prompt
  #
  lane :complete_code_freeze do |options|
```

But again, that looks odd in the IDE (like VSCode) as it ends up appearing as a mix of colors (the color for string literals for the `desc` line — e.g. red in my IDE — but then the color for comments for the doc details — green in my IDE — and that does not look super nicer either when trying to skim the whole file to immediately spot new lanes boundaries…

Which is why overall I think it's nbd after all to sometimes have the tagline of the doc-comment repeated in the `desc` (especially given that it's not always the case depending on how descriptive we want to be in the lane doc-comment), especially given the nicer rendering in most IDEs which make it a delight to skim and navigate the file day to day.

</details>

Overall given the rationales in alternatives above, the format I ended up in the PR is still less noisy (than all those `####…` we have been using before), and still brings a nice rendering in IDEs and easy way of immediately spotting the lane documentation by it all being in the comments (e.g. green) color in its entirety imho.

I encourage you to check how the various options/alternatives look like in your IDE to give me your feedback on which one you prefer and why 🙂

mokagio replied:
I like where this is going and I think the current option is the best of the three you propose.

> Moving all the docs into the `desc` string

+1 on it looking odd as a multiline string with YARD-style docs inside

> Mixing `desc` tagline and the more-detailed doc-comments

> overall I think it's nbd after all to sometimes have the tagline of the doc-comment repeated in the desc

+1, it also moves `desc` away from `lane` but the two are connected. Technically, it's the same thing. But somehow I feel weird about it. It's like the `desc` is for the comment not the `lane`.

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,225 @@
+  # Executes the initial steps needed during code freeze
+  # - Cuts a new release branch
+  # - Extract the Release Notes
+  # - Freezes the GitHub milestone and enable GitHub branch protection for the new branch
```
Just a punctuation and verb tense consistency nitpick:

```suggestion
  # Executes the initial steps needed during code freeze:
  # - Cuts a new release branch
  # - Extracts the Release Notes
  # - Freezes the GitHub milestone and enable GitHub branch protection for the new branch
```

Alternatively:

```suggestion
  # Executes the initial steps needed during code freeze:
  #
  # - Cuts a new release branch
  # - Extracts the Release Notes
  # - Freezes the GitHub milestone and enable GitHub branch protection for the new branch
```

I prefer the new line before the list starts, but I appreciate the value in keeping the comment compact.

mokagio commented on fastlane/Fastfile:
```diff
@@ -38,20 +39,40 @@ EXTENSIONS_BUNDLE_IDENTIFIERS = %w[
 
+
+########################################################################
+# Imports of other Fastfiles
```
Semantics nitpick: I'm unsure about "other Fastfiles" to describe the `lanes/*.rb` files 🤔 

How do you feel about the following?

```suggestion
# Import domain specific lanes
```

mokagio reviewed (commented):
Shaping up good 👍

AliSoftware commented:
ℹ️  PR now rebased on top of the last-minute changes recently made in parent branch during [151657b](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061/commits/151657b761691fc149a768014468b96c01bd6738) (cf https://github.com/wordpress-mobile/WordPress-iOS/pull/18061#issuecomment-1069562053).

_FYI, during that rebase, I amended the "Move all L10n lanes to lanes/localization.rb" commit to update [the call site of `ios_merge_strings_files` here](https://github.com/wordpress-mobile/WordPress-iOS/pull/18140/commits/83ff985f3a2c11baec1c4eeb6e7e2d25e594bc69#diff-ad77c85d7eb3034dec610211cb1747f7cac134e425448b1dcc246d71df3e8639R96-R99) and [the call site of `ios_extract_keys_from_strings_files` here](https://github.com/wordpress-mobile/WordPress-iOS/pull/18140/commits/83ff985f3a2c11baec1c4eeb6e7e2d25e594bc69#diff-ad77c85d7eb3034dec610211cb1747f7cac134e425448b1dcc246d71df3e8639R226-R230)._

AliSoftware committed:
Update rubocop rules

AliSoftware committed:
Fix typos & nits

Applying suggestions from @mokagio

AliSoftware committed:
Add remaining doc-comments for lanes/release.rb

AliSoftware committed:
Doc-comments for `lanes/build.rb`

AliSoftware committed:
Doc-comments for `lanes/codesign.rb`

AliSoftware committed:
Doc-comments for `lanes/localization.rb`

AliSoftware committed:
Doc-comments for `lanes/screenshots.rb`

AliSoftware committed:
rubocop --auto-correct --only 'Layout/EmptyLinesAroundBlockBody'

AliSoftware committed:
rubocop --auto-correct --only 'Style/StringLiterals'

AliSoftware committed:
rubocop --auto-correct --only Layout/SpaceInsideHashLiteralBraces

AliSoftware committed:
rubocop: refine some exceptions

AliSoftware committed:
rubocop --auto-correct --only Style/TrailingCommaInArrayLiteral

AliSoftware committed:
rubocop --auto-correct --only Style/TrailingCommaInArguments

AliSoftware committed:
rubocop --auto-correct --only Layout/SpaceAroundBlockParameters

AliSoftware committed:
rubocop --auto-correct --only Lint/UnusedBlockArgument

AliSoftware committed:
rubocop --auto-correct --only Style/BlockDelimiters

AliSoftware committed:
rubocop --auto-correct --only Layout/MultilineMethodCallBraceLayout

AliSoftware committed:
rubocop --auto-correct --only Style/NumericPredicate

AliSoftware committed:
Use `%w[…]` instead of `'…'.split(' ')`

AliSoftware committed:
rubocop --auto-correct --only Layout/SpaceAfterComma

AliSoftware committed:
rubocop --auto-correct --only Style/TrailingCommaInHashLiteral

AliSoftware committed:
rubocop --auto-correct --only Layout/EmptyLinesAroundMethodBody

AliSoftware committed:
rubocop --auto-correct --only Style/FrozenStringLiteralComment

AliSoftware committed:
rubocop: add exception for Metrics/MethodLength

AliSoftware committed:
Remove extra comma

AliSoftware committed:
rubocop --auto-correct --only Layout/ExtraSpacing

AliSoftware committed:
rubocop: fix Lint/UselessAssignment

AliSoftware committed:
rubocop --auto-correct --only Style/RedundantReturn

AliSoftware committed:
FIx Style/IfUnlessModifier

AliSoftware committed:
rubocop --auto-correct --only Style/StringConcatenation

AliSoftware committed:
rubocop -A --only Style/MutableConstant

AliSoftware committed:
rubocop --auto-correct --only Style/NonNilCheck

AliSoftware committed:
rubocop --auto-correct --only Style/NegatedIf

AliSoftware commented:
@mokagio This PR is now finally ready for review.

I've rebased it on top of `release/19.5` and retargeted the PR to be merged in the release branch, since that's the most likely place from which you'll run it next, which would be a good occasion to validate the changes as well once they land.

I've also updated the PR description with some review guidance, and especially I've provided a small script that should help you review of the PR — to allow you to easily make diffs of the implementation of each lane before/after the move… despite those now having been moved around in different files (making the diff shown in the PR not as useful). HTH!

AliSoftware commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+  # @option [Boolean] create_gh_release If tru, creates a GitHub Release draft after the upload, with zipped xcarchive as artefact
+  # @option [Boolean] beta_release If true, the GitHub release will be marked as being a pre-release
+  #
+  # @called_by build_and_upload_app_store_connect
+  #
+  desc 'Builds and uploads WordPress to TestFlight for distribution'
+  lane :build_and_upload_itc do |options|
```
Note that while doing that big spring cleaning here, that allowed me to notice that `build_and_upload_itc` lane is in fact:
 - Only called by the `build_and_upload_app_store_connect` lane
 - Not called by neither the CI nor usually manually
 - Re-doing the same `ios_build_prechecks` and `ios_build_preflight` calls than its `build_and_upload_app_store_connect` called [already did](https://github.com/wordpress-mobile/WordPress-iOS/pull/18140/files#diff-72b7ec258b7b79c415642e3cb618b99665e0d8e491e4b8aafe468a88636c486fR83-R89)

So this is more like a private lane or helper… used by only one lane… which makes it a bit useless, and tbh is quite confusing as it exposes a new public lane to the user making it unclear if there's a difference between the 2 lanes when trying to figure out which one to call.

I'm thus suggesting that we cut/paste the body/implementation of this lane directly into `build_and_upload_app_store`, then remove this lane (called by no one else anyway), to avoid confusion between the 2 lanes. I've kept this idea for a separate PR though, to avoid making the diff and review of this PR harder than it already is, but hope to follow-up on it once this PR lands.

mokagio replied:
Makes sense 👍

AliSoftware replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/18192

AliSoftware commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+  # @option [Boolean] skip_confirm (default: false) If true, avoids any interactive prompt
+  # @option [Boolean] skip_prechecks (default: false) If true, don't run the ios_build_prechecks and ios_build_preflight
+  #
+  # @called_by build_and_upload_app_center
+  #
+  desc 'Builds and uploads for distribution'
+  lane :build_and_upload_internal do |options|
```
Note that while doing that big spring cleaning here, that allowed me to notice that `build_and_upload_internal` lane is in fact:
 - Only called by the `build_and_upload_app_center` lane
 - Not called by neither the CI nor usually manually
 - Re-doing the same `ios_build_prechecks` and `ios_build_preflight` calls than its `build_and_upload_app_center` called [already did](https://github.com/wordpress-mobile/WordPress-iOS/pull/18140/files#diff-72b7ec258b7b79c415642e3cb618b99665e0d8e491e4b8aafe468a88636c486fR140-R146)

So this is more like a private lane or helper… used by only one lane… which makes it a bit useless, and tbh is quite confusing as it exposes a new public lane to the user making it unclear if there's a difference between the 2 lanes when trying to figure out which one to call.

I'm thus suggesting that we cut/paste the body/implementation of this lane directly into `build_and_upload_app_center`, then remove this lane (called by no one else anyway), to avoid confusion between the 2 lanes. I've kept this idea for a separate PR though, to avoid making the diff and review of this PR harder than it already is, but hope to follow-up on it once this PR lands.

mokagio replied:
Makes sense 👍

AliSoftware replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/18192

AliSoftware commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,236 @@
+  lane :new_beta_release do |options|
+    ios_betabuild_prechecks(options)
+    download_localized_strings_and_metadata(options)
+    # FIXME: (2021.06.17) This is disabled because we currently have a >256 chars string which GlotPress truncates when exporting  the `.strings` files,
+    #   leading to incorrect key for it and (rightful) linter failure. We need to split that key into 2 smaller copies before we can re-enable this.
+    # ios_lint_localizations(input_dir: 'WordPress/Resources', allow_retry: true)
```
👆 _Note to self: remember to re-enable this very soon, after updating the problematic `NSLocalizedString` call in the code — now that we support arbitrary keys which don't have to be the English-copy (via `NSLocalizedString(key, value: "english copy", …)`) and can thus avoid this issue._

AliSoftware commented on fastlane/lanes/screenshots.rb:
```diff
@@ -0,0 +1,230 @@
+# Constants
+#################################################
+
+# rubocop:disable Style/WordArray
+GLOTPRESS_TO_ASC_METADATA_LOCALE_CODES = [
+  ['ar', 'ar-SA'],
+  ['en-gb', 'en-US'],
+  ['en-gb', 'en-GB'],
```
Disabled the rubocop rule because it would have felt a bit strange to write each entry in this map as `%w[ar ar-SA]`, given the real purpose is for this to be a mapping.

We will address this when we will reimplement the `gp_downloadmetadata` action during this sprint (see paaHJt-31O-p2#comment-5845) anyway; especially the new implementation will not only likely require a `Hash` instead of an `Array<Array<String>>`, but might also make use of [the future `LocaleHelper`](https://github.com/wordpress-mobile/release-toolkit/pull/296/files) at call site to provide the locales list in a more readable way. Which means this constant will be revisited soon anyway.

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+      derived_data_path: DERIVED_DATA_PATH,
+      build_for_testing: true,
+      deployment_target_version: options[:ios_version]
+    )
```
I was tempted to suggest DRYing this `run_tests` call and the one for WordPress above because they have a number of common parameters.

But I think it's okay like this. The `run_tests` API shouldn't change much, and there's something useful in having call with all the parameters inline. ⚖️

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+# Lanes related to Building and Testing the code
+#
+platform :ios do
+  # Builds the WordPress app for Testing
+  #
```
I noticed that `build_for_testing` and `build_jetpack_for_testing` use `options` but they don't have documentation for them, while the other lanes in this file do. 

Is it intentional?

AliSoftware replied:
It is not, if options are used they should be documented and if no options are used we should remove the `|options|`; I'll take a look and follow up on this

AliSoftware replied:
Fixed by 2e3ecc73ff1ff45ea9ee4d7e8396306e39ade25e

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+platform :ios do
+  # Builds the WordPress app for Testing
+  #
+  # @called_by CI
```
What's the rationale behind this annotation across various lanes?

Nothing stops a local user from calling this lane, right? And it should work regardless of whether it runs on a developer's machine or in CI.

AliSoftware replied:
The intent was mostly to identify lanes that are called by CI vs lanes that we typically run from our machines, in order to help have a better mental model of what calls what (eg code freeze is called locally, triggers CI, then CI is the one typically calling the build lane).

It was mostly helping me make sense of which lanes typically serves which purpose and which lane was calling which other lane while I was documenting them. Which helped me notice the uselessness of `build_and_upload_itc` above for example. And then I figured since I was trying to make sense of all that call tree on my notepad… I might as well document it with those `@called_by other_lane` annotations to help users figure out which were the "root lanes" vs which were secondary ones. Then `@called_by CI` came naturally when documenting when the caller was not another lane but a `.sh` called by `.buildkite/pipeline.yml`

Ultimately I agree that the fact that a lane is called by CI via pipeline.yml does not mean we can't call it manually too if we want (that was kind of implied from it being a public lane and not a `private_lane` or `def`). But I also thought that leaving this in the doc comment could signal that if we ever change the name of the lane or its expected options we should make sure to update the call made by CI in the `.buildkite/*/*.sh` too 🤷‍♂️

TL;DR: I would not mind if you prefer that I remove those, but figured they wouldn't hurt and could be a useful indicator, even if that isn't intended to mean that we can't call them manually too.

mokagio replied:
> figured they wouldn't hurt and could be a useful indicator, even if that isn't intended to mean that we can't call them manually too.

Okay, fair. I worry they might get out of date, like all comments tend to do. Then again, ideally, we'll call _more_ lanes from CI, not less. Maybe we'll get to a point where it's only useful to highlight a few outlier lanes mean for local dev, and this annotation will become redundant. Till then, okay to keep.

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+    # Find the referenced .xctestrun file based on its name
+    build_products_path = File.join(DERIVED_DATA_PATH, 'Build', 'Products')
+
+    test_plan_path = Dir.glob(File.join(build_products_path, '*.xctestrun')).select do |e|
+      e.include?(options[:name])
```
I'm not sure what `e` stands for? 🤔 

What do you think of this? 

```suggestion
    test_plan_path = Dir.glob(File.join(build_products_path, '*.xctestrun')).select do |path|
      path.include?(options[:name])
```

AliSoftware replied:
Sounds good to me.
Didn't want to change the implementations too much to not make the review of this PR harder that it already is, and make my script producing the lane diffs make minimalistic diffs for easier review; but since you suggested the change here and thus will be aware about it when reviewing the diff.txt files I guess I can commit the suggestion right away 🙂

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+  #
+  # Requires a prebuilt xctestrun file and simulator destination where the tests will be run.
+  #
+  # @option [String] name The (partial) name of the `*.xctestrun` file to run
```
I haven't seen "partial" used in the context of file names. 

There's various options to pick from in [this SO question](https://stackoverflow.com/questions/39007908/filename-without-extension-terminology), like "stem" (_TIL_), but I'd like to suggest this which I think encapsulates the code's expectation

```suggestion
  # @option [String] name The name of the `.xctestrun` file to run, without the `.xctestrun` extension
```

AliSoftware replied:
I think the original documentation used "partial" because it does not have to be the full base name either. For example, `"foo"` can be used to match `WordPressFooTests.xctestplan` here (which your suggestion would miss to document), since we're using `path.include?(…)`, not `File.basename(path) == …` 😉

mokagio replied:
I see, good point... I can't suggest anything that feels better at explaining the parameter without becoming verbose. So 🤷‍♂️

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+
+    sentry_upload_dsym(
+      auth_token: get_required_env('SENTRY_AUTH_TOKEN'),
+      org_slug: 'a8c',
```
Worth extracting this in a constant?

AliSoftware replied:
Addressed in 9fe4717e11614201e84b39bbfbb5aedbd6f1cf17.

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+    )
+  end
+
+  # Builds the WordPress app for an Installable Build ("WordPress Alpha" scheme), and uploads it to AppCenter
```
```suggestion
  # Builds the WordPress app for an Installable Build ("WordPress Alpha" scheme), and uploads it to App Center
```

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+  #
+  desc 'Builds and uploads Jetpack to TestFlight for distribution'
+  lane :build_and_upload_jetpack_for_app_store do
+    jetpack_appstore_code_signing
```
I think it would be useful to run the Sentry CLI check here, too, like `build_and_upload_installable_build ` does:

```suggestion
    sentry_check_cli_installed

    jetpack_appstore_code_signing
```

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+      output_name: 'WordPress Alpha',
+      derived_data_path: DERIVED_DATA_PATH,
+      export_team_id: ENV['INT_EXPORT_TEAM_ID'],
+      export_options: { method: 'enterprise' }
```
🤔 the `export_method` and `export_options` values seemed redundant but [this test](https://github.com/fastlane/fastlane/blob/e874a47c6e2e0e61590a03d3b71e75e5a505d1ce/gym/spec/error_handler_spec.rb#L58-L86) in Fastlane suggests ti might actually need or use both. 

---

How do you feel about groping the values for slightly improved local reasoning?

```suggestion
      clean: true,
      output_directory: BUILD_PRODUCTS_PATH,
      output_name: 'WordPress Alpha',
      derived_data_path: DERIVED_DATA_PATH,
      export_team_id: ENV['INT_EXPORT_TEAM_ID'],
      export_method: 'enterprise',
      export_options: { method: 'enterprise' }
```

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+    UI.message("Successfully built and uploaded installable build here: #{download_url}")
+    install_url = 'https://install.appcenter.ms/orgs/automattic/apps/WPiOS-One-Offs/'
+
+    comment_body = "You can test the <strong>WordPress</strong> changes on this Pull Request by downloading it from AppCenter <a href='#{install_url}'>here</a> with build number: <code>#{build_number}</code>. IPA is available <a href='#{download_url}'>here</a>. If you need access to this, you can ask a maintainer to add you."
```
```suggestion
    comment_body = "You can test the <strong>WordPress</strong> changes on this Pull Request by downloading it from App Center <a href='#{install_url}'>here</a> with build number: <code>#{build_number}</code>. IPA is available <a href='#{download_url}'>here</a>. If you need access to this, you can ask a maintainer to add you."
```

mokagio replied:
Actually, while we're at it, I think this might be clearer:

```suggestion
    comment_body = "You can test the changes in <strong>WordPress</strong> from this Pull Request by downloading it from AppCenter <a href='#{install_url}'>here</a> with build number: <code>#{build_number}</code>. IPA is available <a href='#{download_url}'>here</a>. If you need access to this, you can ask a maintainer to add you."
```

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+      output_name: 'Jetpack Alpha',
+      derived_data_path: DERIVED_DATA_PATH,
+      export_team_id: ENV['INT_EXPORT_TEAM_ID'],
+      export_options: { method: 'enterprise' }
```
Same suggestion as the WordPress version, for consistency:

```suggestion
      configuration: 'Release-Alpha',
      clean: true,
      output_directory: BUILD_PRODUCTS_PATH,
      output_name: 'Jetpack Alpha',
      derived_data_path: DERIVED_DATA_PATH,
      export_team_id: ENV['INT_EXPORT_TEAM_ID'],
      export_method: 'enterprise',
      export_options: { method: 'enterprise' }
```

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+
+    appcenter_upload(
+      api_token: get_required_env('APPCENTER_API_TOKEN'),
+      owner_name: 'automattic',
```
Similar to the Sentry `org_slug` suggestion, would it be worth to extract this in a constant?

AliSoftware replied:
Addressed in 9fe4717e11614201e84b39bbfbb5aedbd6f1cf17.

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+    UI.message("Successfully built and uploaded installable build here: #{download_url}")
+    install_url = 'https://install.appcenter.ms/orgs/automattic/apps/jetpack-installable-builds/'
+
+    comment_body = "You can test the <strong>Jetpack</strong> changes on this Pull Request by downloading it from AppCenter <a href='#{install_url}'>here</a> with build number: <code>#{build_number}</code>. IPA is available <a href='#{download_url}'>here</a>. If you need access to this, you can ask a maintainer to add you."
```
Same suggestion as the WordPress lane.

```suggestion
    comment_body = "You can test the changes in <strong>Jetpack</strong> from this Pull Request by downloading it from App Center <a href='#{install_url}'>here</a> with build number: <code>#{build_number}</code>. IPA is available <a href='#{download_url}'>here</a>. If you need access to this, you can ask a maintainer to add you."
```

Actually, I wonder if it would be best to extract this in a little helper function. I'd say we'd always want the comments to have the same format, so using a template for them might be handy.

This is of course a nitpick, and something that could very well be a followup PR.

AliSoftware replied:
Will DRY the comment text in a follow-up PR

AliSoftware replied:
Done in https://github.com/wordpress-mobile/WordPress-iOS/pull/18192

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -0,0 +1,427 @@
+      dsym_path: lane_context[SharedValues::DSYM_OUTPUT_PATH]
+    )
+
+    if options[:create_release]
```
Note for the [future cleanup](https://github.com/wordpress-mobile/WordPress-iOS/pull/18140/files#r832562218) of this code. IMHO, it would be nice to use the early return style here, i.e. `return unless options[:create_release]`.

AliSoftware replied:
Will tackle that in follow-up PR

AliSoftware replied:
Done in https://github.com/wordpress-mobile/WordPress-iOS/pull/18192

mokagio commented on fastlane/lanes/codesign.rb:
```diff
@@ -0,0 +1,117 @@
+# Lanes related to Code Signing and Provisioning Profiles
+#
+platform :ios do
+  # Registers a new device in the App Store Portal and update all the Provisioning Profiles
```
I _think_ we're referring to the Developer Portal here.

```suggestion
  # Registers a new device in the Developer Portal and update all the Provisioning Profiles
```

mokagio commented on fastlane/Fastfile:
```diff
@@ -38,20 +39,42 @@ EXTENSIONS_BUNDLE_IDENTIFIERS = %w[
+Dotenv.load(USER_ENV_FILE_PATH)
+Dotenv.load(PROJECT_ENV_FILE_PATH)
+GHHELPER_REPO = 'wordpress-mobile/wordpress-iOS'
+ENV['PROJECT_NAME'] = 'WordPress'
+ENV['PUBLIC_CONFIG_FILE'] = File.join(PROJECT_ROOT_FOLDER, 'config', 'Version.Public.xcconfig')
+ENV['INTERNAL_CONFIG_FILE'] = File.join(PROJECT_ROOT_FOLDER, 'config', 'Version.internal.xcconfig')
+ENV['DOWNLOAD_METADATA'] = './fastlane/download_metadata.swift'
+ENV['PROJECT_ROOT_FOLDER'] = "#{PROJECT_ROOT_FOLDER}/"
+ENV['APP_STORE_STRINGS_FILE_NAME'] = 'AppStoreStrings.po'
```
Totally out of scope for this PR, but just saying that I'm looking forward to when we'll be able to drop all these 😄 _One day..._

mokagio commented on fastlane/Fastfile:
```diff
@@ -38,20 +39,42 @@ EXTENSIONS_BUNDLE_IDENTIFIERS = %w[
+ENV['DOWNLOAD_METADATA'] = './fastlane/download_metadata.swift'
+ENV['PROJECT_ROOT_FOLDER'] = "#{PROJECT_ROOT_FOLDER}/"
+ENV['APP_STORE_STRINGS_FILE_NAME'] = 'AppStoreStrings.po'
+ENV['FL_RELEASE_TOOLKIT_DEFAULT_BRANCH'] = 'trunk'
```
```suggestion
```

We should be able to safely remove this export now that it's the default in the release toolkit. In fact, I wonder why I forgot about doing it when upgrading 😞

mokagio commented on fastlane/Fastfile:
```diff
@@ -38,20 +39,42 @@ EXTENSIONS_BUNDLE_IDENTIFIERS = %w[
+# Use this instead of getting values from ENV directly. It will throw an error if the requested value is missing
 def get_required_env(key)
-  unless ENV.key?(key)
-    UI.user_error!("Environment variable '#{key}' is not set. Have you setup #{USER_ENV_FILE_PATH} correctly?")
-  end
+  UI.user_error!("Environment variable '#{key}' is not set. Have you setup #{USER_ENV_FILE_PATH} correctly?") unless ENV.key?(key)
   ENV[key]
 end
```
This method is also used in the lanes we import below. I wonder if it would be more appropriate to define it in a `shared.rb` file that the lanes themselves import but that this one doesn't know about?

Although... maybe it's too much effort. At the end of the day, the lanes ~~will~~ should always and only be called via this `Fastfile`, so the result is the same.

AliSoftware replied:
Yeah I kept it here because of its relationship with `USER_ENV_FILE_PATH`, I say it's not yet worth it to move it until we do some more cleanup at a later stage and have more things to share.

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,236 @@
+  #
+  # - Cuts a new release branch
+  # - Extracts the Release Notes
+  # - Freezes the GitHub milestone and enable GitHub branch protection for the new branch
```
```suggestion
  # - Freezes the GitHub milestone and enables the GitHub branch protection for the new branch
```

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,236 @@
+# Lanes related to the Release Process (Code Freeze, Betas, Final Build, AppStore Submission…)
+#
+platform :ios do
+  # Executes the initial steps needed during code freeze
```
What do you think of 

```suggestion
  # Executes the initial steps of the code freeze
```

or

```suggestion
  # Executes the initial code freeze steps
```

?

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,236 @@
+    print_release_notes_reminder
+  end
+
+  # Completes the final steps for the code freeze
```
For consistency with the wording from the `code_freeze` lane:

```suggestion
  # Executes the final steps for the code freeze
```

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,236 @@
+
+  # Completes the final steps for the code freeze
+  #
+  #  - Generate `.strings` files from code then merge the other, manually-maintained `.strings` files with it
+  #  - Trigger the build of the first beta on CI
```
Again for consistency with the `code_freeze` lane documentation:

```suggestion
  #  - Generates `.strings` files from code then merges the other, manually-maintained `.strings` files with it
  #  - Triggers the build of the first beta on CI
```

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,236 @@
+  # @option [Boolean] skip_confirm (default: false) If true, avoids any interactive prompt
+  # @option [String] version (required) The version number to use for the hotfix (`"x.y.z"`)
+  #
+  desc 'Creates a new hotfix branch for the given `version:x.y.z``. The branch will be cut from the tag `x.y` of the previous release'
```
```suggestion
  desc 'Creates a new hotfix branch for the given `version:x.y.z`. The branch will be cut from the `x.y` tag.'
```

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,236 @@
+    trigger_release_build(branch_to_build: "release/#{version}")
+  end
+
+  # Finalizes a release at the end of a sprint to submit to the AppStore
```
```suggestion
  # Finalizes a release at the end of a sprint to submit to the App Store
```

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,236 @@
+
+  # Finalizes a release at the end of a sprint to submit to the AppStore
+  #
+  #  - updates store metadata
+  #  - bumps final version number
+  #  - removes branch protection and close milestone
+  #  - triggers the final release on CI
```
For consistency with the formatting in the other lanes:

```suggestion
  #  - Updates store metadata
  #  - Bumps final version number
  #  - Removes branch protection and close milestone
  #  - Triggers the final release on CI
```

mokagio commented on fastlane/lanes/release.rb:
```diff
@@ -0,0 +1,236 @@
+  paths[app.to_sym] || UI.user_error!("Invalid app name passed to lane: #{app}")
+end
+
+# Prints the message to remind the Release Manager to audit the Release Notes after code freeze.
```
I'm probably the one who wrote this in the first place, but here's a suggestion to make it more succinct:

```suggestion
# Prints a reminder to audit the Release Notes after code freeze.
```

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -0,0 +1,230 @@
+      end
+    end
+
+    puts languages
```
Maybe this could become a `UI.message`?

AliSoftware replied:
Addressed in 1724d4d04527a3dc69ccebf5476db884aa623ce5

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -0,0 +1,230 @@
+    end
+  end
+
+  # Generates the localized raw screenshots for Jetpack, using the dedicated `"JetpackScreenshotGeneration"` UI test scheme
```
Nitpick: The `""` seem unnecessary when the scheme name is already within backticks.

```suggestion
  # Generates the localized raw screenshots for Jetpack, using the dedicated `JetpackScreenshotGeneration` UI test scheme
```

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -0,0 +1,230 @@
+  #
+  desc 'Creates promo screenshots'
+  lane :create_promo_screenshots do |options|
+    # Run screenshots generator tool
+    # All file paths are relative to the `fast file`.
```
```suggestion
    # All file paths are relative to the `Fastfile`.
```

mokagio commented on fastlane/lanes/screenshots.rb:
```diff
@@ -0,0 +1,230 @@
+  #
+  desc 'Creates Jetpack promo screenshots'
+  lane :create_jetpack_promo_screenshots do |options|
+    # Run screenshots generator tool
+    # All file paths are relative to the `fast file`.
```
```suggestion
    # All file paths are relative to the `Fastfile`.
```

AliSoftware committed:
Add missing @options doc to 2 lanes

AliSoftware committed:
DRY some common values in constants

See https://github.com/wordpress-mobile/WordPress-iOS/pull/18140/files#r832794330

AliSoftware committed:
Fix doc: `create_release` vs `create_gh_release`

AliSoftware committed:
Use UI.message instead of puts in screenshots lane

AliSoftware committed:
Apply suggestions from code review

Thanks a lot @mokagio for the great wording suggestions and eagle eye at spotting inconsistencies in doc comments!

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

AliSoftware commented:
@mokagio Thank you so much for your detailed review, especially on the lane doc-comments and wording suggestions!

I think I addressed all of your feedback.
 - For most of your GitHub suggestions, I've applied them directly in this PR
 - For more involved changes, like the `build_and_upload_{internal,itc}` lanes removal, the DRYing of the Installable Builds PR comment and the early return for `options[:create_release]`, I'd addressed them in [this separate PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/18192) which builds on top of this one, to avoid mixing more involved changes like those in this one.

AliSoftware commented:
EDIT: Rebased the whole branch on top of `release/19.5` (which I apparently totally forgot to do earlier when I updated the the PR to target `release/19.5` post code-freeze 😅 ). All should be set now!

mokagio reviewed (approved):
:shipit:

mokagio commented:
I just noticed that Fastlane's shell tab-completion no longer works after this change 😭 

Of course, this change is more valuable than me tab-completing my way through my daily tasks, but I just wanted to note it down for future reference.

The tab-completion code is [minimalistic](https://github.com/fastlane/fastlane/blob/668d976ec61d74509f3123ea1b7fe1836a1052e4/fastlane/lib/assets/completions/completion.zsh). It does not understand the code, simply reads it. It looks for the first Fastfile-like it can find, then looks for `lane` lines and extract their names. But now that we split our lanes into multiple files, that approach no longer works.

I guess one would have to do something more advanced like actually parsing the `Fastfile` using Fastlane's out methods then extract the list of lanes from there.

AliSoftware commented:
Ah, good point (and TIL that completion script existed!).
Well to be fair I guess it didn't work with `Jetpack-Fastfile` either before the split… but yeah.

You peaked my interest, if I find some time I'll see if I can tweak that completion script to parse the lanes in a more resilient way 😛

AliSoftware commented:
> I just noticed that Fastlane's shell tab-completion no longer works after this change 😭

[Your prayers have been answered](https://github.com/wordpress-mobile/WordPress-iOS/pull/18223/files) 😄
